docs/api-configuration.md
# API configuration

Currently, the API is configured entirely with environment variables:

- ### `NODE_ENV`

  - This variable is used to determine which database connection configuration
    to use while the app is running. Knex reads its config information from
    `/api/knexfile.js`, which exports a few different configurations. Knex picks
    the configuration that matches the current `NODE_ENV`.
  - This variable also controls which seed data is applied to the database when
    using the Knex `seed` command.
  - Supported values:
    - `development` - used during development (surprise!). If you're developing
      with Docker, this gets set by the Docker config.
    - `test` - used during testing, eventually. Not used right now.
    - `production` - used in our staging and production environments.
  - **default**: development

- ### `SESSION_SECRET`

  - This is the secret used to sign client authentication cookies. It should be
    cryptographically strong. This secret should be protected.
  - If the secret is changed and the API process restarted, all users will be
    effectively logged out. If the secret is not set and the default value is
    used, users will be logged out every time the process restarts.
  - This value should be at least 256 bits, preferably as a hex string. If the
    value is a hex string, it will be converted to a byte buffer. A final secret
    that is below 256 bits will log a warning.
  - **default**: cryptographically-strong random hex string representing a
    512-byte secret

- ### `SESSION_LIFETIME_MINUTES`

  - How long a user may be logged in without logging in again. Currently, any
    user activity involving the API will reset the session expiration to this
    lifetime. E.g., a user logs in and gets a 2-day session. If they do
    nothing, they will be logged out in 2 days. However, if after 1 day they
    load an APD, their session is extended to 2 more days.
  - **default**: 2880 minutes, or 2 days

- ### `PORT`

  - This is the port that the API process should listen on.
  - **default**: 8000

- ### `DATABASE_URL`

  - This variable is only used if `NODE_ENV` is set to `production`.
  - This is the connection URL for the production environment database. In our
    current deployment setup, the cloud provider sets this for us. If it has to
    be set manually, it must be something that the underlying database client
    understands.
  - **default**: none

- ### `DEV_DB_HOST`

  - This variable is only used if `NODE_ENV` is set to `development`.
  - If you are developing the API code and _not_ using Docker, the default
    development database configuration won't work for you because it expects to
    use a database host that is provided by Docker. To get around that, set this
    variable to the hostname of your local PostgreSQL instance.
  - **default**: none

- ### `TEST_DB_HOST`

  - This variable is only used if `NODE_ENV` is set to `test`.
  - If you are running tests and _not_ using Docker, the default test database
    configuration won't work for you because it expects to use a database host
    that is provided by Docker. To get around that, set this variable to the
    hostname of your local PostgreSQL instance.
  - **default**: none

# File store

The API also supports file storage to different storage locations, depending on
configuration. The following environment variables determine which blob store
we use:

- ### `FILE_STORE`

  - This determines which type of store to use. For invalid values, defaults
    to `local`.
    Valid vlues are:
    - `local` - files are stored on the local filesystem
    - `s3` - files are stored in an S3 bucket

- ### `FILE_PATH`

  - Required if `FILE_STORE` is set to `local`. This is the file path for
    reading and writing files. Can either be absolute or relative to the
    API server's working directory. Defaults to `__files` relative to the
    API's working directory.

- ### `FILE_S3_BUCKET`

  - Required if `FILE_STORE` is set to `s3`. This is the bucket name for
    reading and writing files. It is _only_ the bucket name, not an S3 URI.
    E.g., `magic-bucket.the-app.gov`, not `s3://magic-bucket.the-app.gov`

- ### `AWS_ACCESS_KEY_ID`

  - Required if `FILE_STORE` is set to `s3`. AWS access key ID.

- ### `AWS_SECRET_ACCESS_KEY`

  - Required if `FILE_STORE` is set to `s3`. AWS secret access key.


CHANGELOG.md
# Next release

Anticipated release: January 6, 2019

#### 🚀 New features

#### 🐛 Bugs fixed

- Fixed a bug where an admin could not edit a user's account without changing that user's email address ([#1973])

#### ⚙️ Behind the scenes

# 2.2.0

Released: December 18, 2019

#### 🚀 New features

- Make disabled button state more explicit ([#1957])
- Upload files from the rich text editor instead of embedding them directly into the text ([#1735])
- Made the rich text areas resize to fit their contents
- Automatically save APDs as changes are made
- Removed the 11 Standards and Conditions text fields in favor of two simpler fields ([#1947])

#### 🐛 Bugs fixed

- Fixed an issue with the sidebar scrolling into the footer ([#1967])

#### ⚙️ Behind the scenes

- Added OWASP ZAP active security scan in continuous integration and deployment ([#1928])

# 2.1.0

Released: December 3, 2019

#### 🚀 New features

- Added a print preview page ([#1921])
- Display a loading screen while waiting for the APD to load so users can tell the app is doing something. ([#1368])

#### 🐛 Bugs fixed

- Fixed a bug where contractor hourly rates were not used in budget calculations ([#1925])
- Switches to a new rich text editor that is better supported ([#1871])

#### ⚙️ Behind the scenes

- Removed unused code ([#1801], [#1802], [#1803])

# 2.0.1

Released: November 18, 2019

#### 🐛 Bugs fixed

- Updated dependencies with known vulnerabilities

# 2.0.0

Released: November 18, 2019

#### 🚀 New features

- Alert users on yellow support browsers that they should upgrade ([#1904])
- Don't try to load the page on unsupported browsers ([#1882])

#### 🐛 Bugs fixed

- Fixed an issue where the label for the state name dropdown in the State Profile section was not correctly associated ([#1779])
- Fixed an issue where some table header cells were empty ([#1780])
- Fixed an issue where the account management and logout links in the header were unresponsive from the APD page ([#1914])

#### ⚙️ Behind the scenes

- Switched from a relational data model for APDs to a document-oriented model ([#1793], [#1794], [#1796], [#1798], [#1800], [#1826])
- Removed unused code ([#1799])

# v1.1.0

Released: September 27, 2019

#### 🚀 New features

- Attempt to alert users before automatically logging them out due to inactivity. The app will try to use built-in browser notifications as well as flashing the tab title. In browsers that support it, the app treats activity in any eAPD tab as valid, so all eAPD tabs will remain valid as long as at least one of them is getting activity. ([#1697])
- Scroll the collapsed activity review panel into view after collapsing an activity form. This way, when you collapse an activity, you end up essentially looking at the list of activities again instead of being pushed way down the page. ([#1732])
- Automatically select numeric form field contents when the field is focused if the current value is 0 ([#1736])
- Add spellcheck to rich text fields ([#1769])

#### 🐛 Bugs fixed

- Add a "skip to main content" link ([#1304])
- Prompt for confirmation before deleting APD key personnel ([#1651], [#1647])
- Fixed accessibility issues on the login page and the dashboard ([#1688])
- Fixed semantic heading levels ([#1695])
- Fixed a keyboard focus order problem when adding new items to a list ([#1712])
- Fixed a keyboard focus order problem with the system use banner ([#1715])
- Fixed the save button so it doesn't go into the footer ([#1524])
- Switched remaining text inputs (except for rich text) to Design System components and removed custom components ([#1686])
- For users in American Samoa, Guam, Northern Mariana Islands, or U.S. Virgin Islands, do not attempt to display a territory outline at the top of the sidebar because we don't have those outlines ([#1423]; see [#1730] for more information)
- Fixed a bug where adding a new contractor resource didn't expand the contractor form ([#1710])
- Make the navigation sidebar scrollable. ([#1475])
- Fixed the width on activity overview detail and statement of alternatives textbox labels ([#1640])
- Made the "manage account" and "login" text the same size in the header dropdown ([#1655], [#1680])
- Removed the emoji button from the rich text editor ([#1649])
- Extend user sessions on API activity ([#1728])
- Adds spacing around the Medicaid director and Medicaid office address form headers ([#1646])
- Adds margin to the bottom of the state dashboard ([#1601])
- Adds an ARIA region component to prevent screen readers from prematurely announcing quarterly budget numbers ([#1731])
- Fixed alignment of the message if there are no APDs on the state dashboard ([#1602])
- Adds spacing between the login form and the "forgotten password" help link ([#1600])
- Adds per-year APD key personnel costs to the review view ([#1747])
- Adds a more informative error message if attempting to save an APD while logged out ([#1729])
- Use a dollar field for the "other funding" field in the cost allocation form ([#1754])
- Fixes a screen reader bug by adding a wrapper around the CMS Design System's Choice component ([#1760])
- Round off dollar input fields when they lose focus ([#1739])
- Fixed a bug where multiline plain text fields (such as the activity overview) gets exported as a text field, and the content is truncated within it. ([#1767])
- Fixed a bug where the session authentication cookie would expire at the end of the browser session instead of at the scheduled time. ([#1756])
- Fixed the spacing in the activity overview section. ([#1648])
- Disabled cacheing of the index page, so that clients always get the latest. ([#1775])
- Fixed a bug where the APD Key Personnel section asked for a person's percent time using a plain number input box instead of a percent box ([#1753])
- Improved ARIA metadata on the account management dropdown button for screen readers ([#1681])
- Fixed a bug where the sidebar would change width during scrolling in IE. ([#1434])

api/env.js
const crypto = require('crypto');
const dotenv = require('dotenv');

const defaults = {
  // Default to locking accounts for 30 minutes if there are five
  // failed logon attempts within a 1-minute period.
  AUTH_LOCK_FAILED_ATTEMPTS_COUNT: 5,
  AUTH_LOCK_FAILED_ATTEMPTS_WINDOW_TIME_MINUTES: 1,
  AUTH_LOCK_FAILED_ATTEMPTS_DURATION_MINUTES: 30,
  FILE_PATH: '__files',
  FILE_STORE: 'local',
  LOG_LEVEL: 'info',
  LOG_FILE: 'false',
  LOG_CONSOLE: 'true',
  NODE_ENV: 'development',
  PORT: 8000,
  PROXY_TRUST: 'false',
  SESSION_LIFETIME_MINUTES: 2880,
  SESSION_SECRET: crypto.randomBytes(64).toString('hex'),
  STORE_PATH: '',
  STORE_TYPE: 'null' // default to using the /dev/null store
};

// Check if there are any CloudFoundry user-provided services
// offering up environment variables.  If there are, extract
// them and merge them into our environment.
const upsEnv = {};
if (process.env.VCAP_SERVICES) {
  try {
    const vcap = JSON.parse(process.env.VCAP_SERVICES);
    if (Array.isArray(vcap['user-provided'])) {
      vcap['user-provided'].forEach(ups => {
        Object.entries(ups.credentials).forEach(([name, value]) => {
          upsEnv[name] = value;
        });
      });
    }
  } catch (e) {} // eslint-disable-line no-empty
}

dotenv.config();
process.env = Object.assign({}, defaults, upsEnv, process.env);

// Don't require this until process.env is finished setting up, since that
// defines how the logger works.
const logger = require('./logger')('env setup');

// Convert the SESSION_SECRET from a hex string to a bunch of bytes, if
// applicable.
if (process.env.SESSION_SECRET.match(/^[a-f0-9]+$/i)) {
  process.env.SESSION_SECRET = Buffer.from(process.env.SESSION_SECRET, 'hex');
} else {
  logger.warn('SESSION_SECRET should be a hex string');
}

if (process.env.SESSION_SECRET.length * 8 < 256) {
  logger.warn(
    `SESSION_SECRET is ${process.env.SESSION_SECRET.length *
      8} bits - should be AT LEAST 256 bits [see: https://tools.ietf.org/html/rfc7518#section-3.2]`
  );
}
